File: src/main/java/com/impact/mods/gregtech/tileentities/multi/units/GTMTE_Aerostat.java
package com.impact.mods.gregtech.tileentities.multi.units;

import com.impact.core.Impact_API;
import com.impact.impact;
import com.impact.mods.gregtech.blocks.Casing_Helper;
import com.impact.mods.gregtech.enums.Texture;
import com.impact.mods.gregtech.tileentities.multi.implement.GTMTE_Impact_BlockBase;
import com.impact.network.IPacketString;
import com.impact.util.PositionObject;
import com.impact.util.Utilits;
import com.impact.util.string.MultiBlockTooltipBuilder;
import com.impact.util.vector.Vector3i;
import com.impact.util.vector.Vector3ic;
import gregtech.api.enums.Materials;
import gregtech.api.enums.Textures;
import gregtech.api.interfaces.ITexture;
import gregtech.api.interfaces.metatileentity.IMetaTileEntity;
import gregtech.api.interfaces.tileentity.IGregTechTileEntity;
import gregtech.api.render.TextureFactory;
import gregtech.api.util.GT_Utility;
import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumChatFormatting;
import net.minecraftforge.common.util.ForgeDirection;
import space.impact.api.multiblocks.structure.IStructureDefinition;
import space.impact.api.multiblocks.structure.StructureDefinition;

import java.util.ArrayList;
import java.util.List;

import static com.impact.mods.gregtech.blocks.Build_Casing_Helper.AEROSTATE_PLATFORM;
import static space.impact.api.multiblocks.structure.StructureUtility.ofBlock;

public class GTMTE_Aerostat extends GTMTE_Impact_BlockBase<GTMTE_Aerostat> implements IPacketString {
	
	public final static int MAX_BUFFER = 100_000;
	static Block CASING = Casing_Helper.sCaseCore3;
	static int CASING_META = AEROSTATE_PLATFORM.getMeta();
	static ITexture INDEX_CASE = Textures.BlockIcons.casingTexturePages[3][CASING_META + 32];
	static int CASING_TEXTURE_ID = AEROSTATE_PLATFORM.getIDCasing();
	static IStructureDefinition<GTMTE_Aerostat> definition =
			StructureDefinition.<GTMTE_Aerostat>builder()
					.addShape("main", new String[][]{
							{"AAA"},
							{"A~A"},
							{"AAA"},
					})
					.addElement('A', ofBlock(CASING, CASING_META))
					.build();
	public int timer = 20;
	public int curID = 0;
	public boolean firstOpen = true;
	public boolean isFullBuffer = false;
	public int curBuffer = 0;
	public String playerName = "";
	public String aerName = "";
	public List<GTMTE_Aerostat> currentLocationPlatforms = new ArrayList<>();
	
	public GTMTE_Aerostat(int aID, String aNameRegional) {
		super(aID, "impact.multis.aerostat", aNameRegional);
	}
	
	public GTMTE_Aerostat(String aName) {
		super(aName);
	}
	
	public static List<GTMTE_Aerostat> getRadiusAeroStates(String owner, IGregTechTileEntity te) {
		List<GTMTE_Aerostat> as = new ArrayList<>();
		for (String name : Impact_API.sAerostat.keySet()) {
			PositionObject location = new PositionObject(Impact_API.sAerostat.get(name));
			PositionObject thisPos = new PositionObject(te);
			IGregTechTileEntity teTarget = te.getIGregTechTileEntity(location.xPos, location.yPos, location.zPos);
			if (teTarget == null) continue;
			if (teTarget.getMetaTileEntity() instanceof GTMTE_Aerostat) {
				GTMTE_Aerostat gtmte_aerostat = (GTMTE_Aerostat) teTarget.getMetaTileEntity();
				if (gtmte_aerostat.playerName.equals(owner)) {
					int distance = Utilits.distanceBetween3D(thisPos.xPos, location.xPos, thisPos.yPos, location.yPos, thisPos.zPos, location.zPos);
					if (distance >= 16 && distance <= 512) {
						as.add(gtmte_aerostat);
					}
				}
			}
		}
		return as;
	}
	
	@Override
	public ITexture[] getTexture(final IGregTechTileEntity aBaseMetaTileEntity, final byte aSide, final byte aFacing, final byte aColorIndex, final boolean aActive, final boolean aRedstone) {
		return aSide == 1 ? new ITexture[]{INDEX_CASE, TextureFactory.of(Texture.Icons.PLATFORM_AEROSTATE_OVERLAY)} : new ITexture[]{INDEX_CASE};
	}
	
	@Override
	public IMetaTileEntity newMetaEntity(IGregTechTileEntity aTileEntity) {
		return new GTMTE_Aerostat(this.mName);
	}
	
	public boolean isFacingValid(byte aFacing) {
		return aFacing > 1;
	}
	
	@Override
	public IStructureDefinition<GTMTE_Aerostat> getStructureDefinition() {
		return definition;
	}
	
	@Override
	public Object getClientGUI(int aID, InventoryPlayer aPlayerInventory, IGregTechTileEntity aBaseMetaTileEntity) {
		return null;
	}
	
	@Override
	public Object getServerGUI(int aID, InventoryPlayer aPlayerInventory, IGregTechTileEntity aBaseMetaTileEntity) {
		return null;
	}
	
	@Override
	public boolean machineStructure(IGregTechTileEntity thisController) {
		mWrench = mScrewdriver = mSoftHammer = mHardHammer = mSolderingTool = mCrowbar = true;
		if (thisController.getYCoord() < 60) {
			return false;
		}
		
		final Vector3ic forgeDirection = new Vector3i(
				ForgeDirection.getOrientation(thisController.getBackFacing()).offsetX,
				ForgeDirection.getOrientation(thisController.getBackFacing()).offsetY,
				ForgeDirection.getOrientation(thisController.getBackFacing()).offsetZ
		);
		
		boolean formationChecklist = true;
		
		for (int X = -1; X <= 1; X++) {
			for (int Z = -1; Z <= 1; Z++) {
				if (X == 0 && Z == 0) {
					continue;
				}
				final Vector3ic offset = rotateOffsetVector(forgeDirection, X, 0, Z);
				IGregTechTileEntity currentTE = thisController.getIGregTechTileEntityOffset(offset.x(), offset.y(), offset.z());
				
				//Utilits.setBlock(thisController, offset.x(), offset.y(), offset.z(), CASING, CASING_META);
				
				if (!super.addInputToMachineList(currentTE, CASING_TEXTURE_ID)) {
					if ((thisController.getBlockOffset(offset.x(), offset.y(), offset.z()) == CASING) && (thisController.getMetaIDOffset(offset.x(), offset.y(), offset.z()) == CASING_META)) {
					} else {
						formationChecklist = false;
					}
				}
			}
		}
		
		if (mInputHatches.size() != 1) {
			formationChecklist = false;
		}
		
		return formationChecklist;
	}
	
	@Override
	public void onPostTick(IGregTechTileEntity aBaseMetaTileEntity, long aTick) {
		super.onPostTick(aBaseMetaTileEntity, aTick);
		if (aBaseMetaTileEntity.isServerSide() && mMachine) {
			if (aTick % 20 == 0 && curBuffer < MAX_BUFFER - 1000) {
				isFullBuffer = false;
			}
			if (!isFullBuffer && (depleteInput(Materials.NatruralGas.getGas(1000)) || depleteInput(Materials.Gas.getGas(1000)))) {
				curBuffer += 1000;
				if (curBuffer + 1000 > MAX_BUFFER) {
					curBuffer    = MAX_BUFFER;
					isFullBuffer = true;
				}
			}
		}
	}
	
	@Override
	protected MultiBlockTooltipBuilder createTooltip() {
		MultiBlockTooltipBuilder b = new MultiBlockTooltipBuilder("aerostate");
		b
				.addInfo("info.0", "Moving over a distance")
				.addInfo("info.1", "Natural gas is used as fuel")
				.addInfo("info.2", "Minimum height (Y coord) for work: 60")
				.addTypeMachine("type", "Moving Machine")
				.addController()
				.sizeStructure(3, 1, 3)
				.addInputHatch(1)
				.addCasingInfo("case.1", "Aerostate Platform Casing", 7)
				.signAndFinalize();
		return b;
	}
	
	@Override
	public boolean onRightclick(IGregTechTileEntity aBaseMetaTileEntity, EntityPlayer aPlayer) {
		return mMachine && super.onRightclick(aBaseMetaTileEntity, aPlayer);
	}
	
	public void setLocationName(String name) {
		try {
			PositionObject thisLocation = new PositionObject(getBaseMetaTileEntity());
			if (!Impact_API.sAerostat.containsKey(name)) {
				Impact_API.sAerostat.remove(aerName);
				aerName = name;
				impact.proxy.addClientSideChatMessages("Set location name: \"" + aerName + "\"");
				Impact_API.sAerostat.put(aerName, thisLocation.getCoords());
				curID = 1;
			} else {
//				Utilits.sendChatByTE(getBaseMetaTileEntity(), "This name is already in use!");
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	@Override
	public void onFirstTick(IGregTechTileEntity te) {
		super.onFirstTick(te);
		if (te.isServerSide()) {
			if (!te.getOwnerName().isEmpty()) {
				playerName = te.getOwnerName();
			}
			if (!aerName.isEmpty()) {
				setLocationName(aerName);
			}
		}
	}
	
	@Override
	public void saveNBTData(NBTTagCompound aNBT) {
		super.saveNBTData(aNBT);
		aNBT.setInteger("timer", timer);
		aNBT.setInteger("curBuffer", curBuffer);
		aNBT.setInteger("curID", curID);
		aNBT.setString("aerName", aerName);
		aNBT.setString("playerName", playerName);
		aNBT.setBoolean("firstOpen", firstOpen);
	}
	
	@Override
	public void inValidate() {
		Impact_API.sAerostat.remove(aerName);
//		if (!aerName.equals("")) {
//			Utilits.sendChatByTE(getBaseMetaTileEntity(), "Remove Station: \"" + aerName + "\"");
//		}
		super.inValidate();
	}
	
	@Override
	public void loadNBTData(NBTTagCompound aNBT) {
		super.loadNBTData(aNBT);
		timer      = aNBT.getInteger("timer");
		curID      = aNBT.getInteger("curID");
		curBuffer  = aNBT.getInteger("curBuffer");
		firstOpen  = aNBT.getBoolean("firstOpen");
		aerName    = aNBT.getString("aerName");
		playerName = aNBT.getString("playerName");
	}
	
	@Override
	public void onNotePadRightClick(byte aSide, EntityPlayer aPlayer, float aX, float aY, float aZ) {
		super.onNotePadRightClick(aSide, aPlayer, aX, aY, aZ);
		if (aPlayer.capabilities.isCreativeMode) {
			for (String name : Impact_API.sAerostat.keySet()) {
				GT_Utility.sendChatToPlayer(aPlayer, "Name: " + EnumChatFormatting.RED + name);
			}
		}
	}
	
	@Override
	public boolean checkRecipe(ItemStack itemStack) {
		return false;
	}
	
	public int getDistanceTravel() {
		int distance = 0;
		int id = 1;
		for (GTMTE_Aerostat aerostat : currentLocationPlatforms) {
			if (id == curID) {
				PositionObject thisPos = new PositionObject(getBaseMetaTileEntity());
				PositionObject newPos = new PositionObject(aerostat.getBaseMetaTileEntity());
				distance = Utilits.distanceBetween3D(thisPos.xPos, newPos.xPos, thisPos.yPos, newPos.yPos, thisPos.zPos, newPos.zPos);
				if (distance >= 16 && thisPos.dPos == newPos.dPos) {
					break;
				}
			}
			id++;
		}
		return distance;
	}
	
	public void toTravel(EntityPlayer aPlayer) {
		
		int id = 1;
		for (GTMTE_Aerostat aerostat : currentLocationPlatforms) {
			if (id == curID) {
				PositionObject thisPos = new PositionObject(getBaseMetaTileEntity());
				PositionObject newPos = new PositionObject(aerostat.getBaseMetaTileEntity());
				int distance = Utilits.distanceBetween3D(thisPos.xPos, newPos.xPos, thisPos.yPos, newPos.yPos, thisPos.zPos, newPos.zPos);
				if (distance >= 16 && thisPos.dPos == newPos.dPos) {
					if (curBuffer - distance * 25 > 0) {
						curBuffer -= distance * 25;
						aPlayer.setPositionAndUpdate(newPos.xPos + 0.5D, newPos.yPos + 1, newPos.zPos + 0.5D);
					} else {
						GT_Utility.sendChatToPlayer(aPlayer, "No Fuel");
					}
				} else {
					GT_Utility.sendChatToPlayer(aPlayer, "Error Distance");
				}
				return;
			}
			id++;
		}
	}
	
	public void onScrewdriverRightClick(byte aSide, EntityPlayer aPlayer, float aX, float aY, float aZ) {
		super.onScrewdriverRightClick(aSide, aPlayer, aX, aY, aZ);
		GT_Utility.sendChatToPlayer(aPlayer, "Owner " + getBaseMetaTileEntity().getOwnerName());
		GT_Utility.sendChatToPlayer(aPlayer, "NameCustom " + playerName);
	}
	
	@Override
	public void construct(ItemStack itemStack, boolean b) {
		buildPiece(itemStack, b, 1, 0, 1);
	}
	
	@Override
	public void update(String... str) {
		setLocationName(str[0]);
	}
	
	@Override
	public boolean hasSeparate() {
		return false;
	}
}

File: src/main/java/com/impact/mods/gregtech/tileentities/multi/units/GTMTE_Aerostat.kt
package com.impact.mods.gregtech.tileentities.multi.units

import com.google.common.io.ByteArrayDataInput
import com.impact.addon.gt.api.aerostat.IAeroStat
import com.impact.addon.gt.api.position.IPosition
import com.impact.common.managers.AeroStateNetworkManager
import com.impact.impact
import com.impact.mods.gregtech.GT_ItemList
import com.impact.mods.gregtech.blocks.Build_Casing_Helper
import com.impact.mods.gregtech.blocks.Casing_Helper
import com.impact.mods.gregtech.enums.Texture
import com.impact.mods.gregtech.tileentities.hatches.GTMTEAerostatCargoHatch
import com.impact.mods.gregtech.tileentities.multi.implement.GTMTE_Impact_BlockBase
import com.impact.util.PositionObject
import com.impact.util.Utilits
import com.impact.util.string.MultiBlockTooltipBuilder
import com.impact.util.vector.Vector3i
import com.impact.util.vector.Vector3ic
import gregtech.api.enums.Materials
import gregtech.api.enums.Textures
import gregtech.api.interfaces.ITexture
import gregtech.api.interfaces.metatileentity.IMetaTileEntity
import gregtech.api.interfaces.tileentity.IGregTechTileEntity
import gregtech.api.metatileentity.implementations.GT_MetaTileEntity_Hatch
import gregtech.api.render.TextureFactory
import gregtech.api.util.GT_Utility
import net.minecraft.block.Block
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.InventoryPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntityFurnace
import net.minecraftforge.common.util.ForgeDirection
import space.impact.api.multiblocks.structure.IStructureDefinition
import space.impact.api.multiblocks.structure.StructureDefinition
import space.impact.api.multiblocks.structure.StructureUtility
import space.impact.packet_network.network.packets.IStreamPacketReceiver
import java.util.*

class GTMTE_Aerostat : GTMTE_Impact_BlockBase<GTMTE_Aerostat>, IStreamPacketReceiver, IAeroStat {

    companion object {
        const val MAX_BUFFER = 50_000
        val CASING: Block = Casing_Helper.sCaseCore3
        val CASING_META: Int = Build_Casing_Helper.AEROSTATE_PLATFORM.meta
        val INDEX_CASE: ITexture = Textures.BlockIcons.casingTexturePages[3][CASING_META + 32]
        val CASING_TEXTURE_ID: Int = Build_Casing_Helper.AEROSTATE_PLATFORM.idCasing
        private val STRUCTURE = StructureDefinition.builder<GTMTE_Aerostat>()
            .addShape(
                "main", arrayOf(
                    arrayOf("AAA"),
                    arrayOf("A~A"),
                    arrayOf("AAA"),
                )
            )
            .addElement('A', StructureUtility.ofBlock(CASING, CASING_META))
            .build()
    }

    constructor(aID: Int, aNameRegional: String?) : super(aID, "impact.multis.aerostat", aNameRegional)
    constructor(aName: String?) : super(aName)

    @JvmField
    var indexSelect: Int = 0

    @JvmField
    var currentLocationPlatforms: MutableList<IAeroStat> = ArrayList()

    @JvmField
    var curBuffer: Int = 0

    @JvmField
    var selectedDistance = 0

    private var firstOpen: Boolean = true
    private var isFullBuffer: Boolean = false
    private var isFluidEngine: Boolean = false
    var targetPos: IPosition? = null
    private var cargoHatch: GTMTEAerostatCargoHatch? = null

    override var owner: String = ""
    override var name: String = ""
    var uuid: String = UUID.randomUUID().toString()

    override fun hasIndicator(): Boolean {
        return true
    }

    override fun getStack(): ItemStack {
        return GT_ItemList.Aerostat.get(1)
    }

    override fun getTexture(aBaseMetaTileEntity: IGregTechTileEntity, aSide: Byte, aFacing: Byte, aColorIndex: Byte, aActive: Boolean, aRedstone: Boolean): Array<ITexture> {
        return if (aSide.toInt() == 1) arrayOf(INDEX_CASE, TextureFactory.of(Texture.Icons.PLATFORM_AEROSTATE_OVERLAY)) else arrayOf(INDEX_CASE)
    }

    override fun newMetaEntity(aTileEntity: IGregTechTileEntity): IMetaTileEntity {
        return GTMTE_Aerostat(this.mName)
    }

    override fun isFacingValid(aFacing: Byte): Boolean {
        return aFacing > 1
    }

    override fun getStructureDefinition(): IStructureDefinition<GTMTE_Aerostat> = STRUCTURE

    override fun getClientGUI(aID: Int, aPlayerInventory: InventoryPlayer, aBaseMetaTileEntity: IGregTechTileEntity): Any? {
        return null
    }

    override fun getServerGUI(aID: Int, aPlayerInventory: InventoryPlayer, aBaseMetaTileEntity: IGregTechTileEntity): Any? {
        return null
    }

    override fun machineStructure(thisController: IGregTechTileEntity): Boolean {
        cargoHatch = null
        mCrowbar = true
        mSolderingTool = mCrowbar
        mHardHammer = mSolderingTool
        mSoftHammer = mHardHammer
        mScrewdriver = mSoftHammer
        mWrench = mScrewdriver
        if (thisController.yCoord < 60) {
            return false
        }

        val forgeDirection: Vector3ic = Vector3i(
            ForgeDirection.getOrientation(thisController.backFacing.toInt()).offsetX,
            ForgeDirection.getOrientation(thisController.backFacing.toInt()).offsetY,
            ForgeDirection.getOrientation(thisController.backFacing.toInt()).offsetZ
        )

        var formationChecklist = true

        for (X in -1..1) {
            for (Z in -1..1) {
                if (X == 0 && Z == 0) {
                    continue
                }
                val offset = rotateOffsetVector(forgeDirection, X, 0, Z)
                val currentTE = thisController.getIGregTechTileEntityOffset(offset.x(), offset.y(), offset.z())

                if (!addInputToMachineList(currentTE, CASING_TEXTURE_ID)) {
                    if ((thisController.getBlockOffset(offset.x(), offset.y(), offset.z()) === CASING)
                        && (thisController.getMetaIDOffset(offset.x(), offset.y(), offset.z()).toInt() == CASING_META)
                    ) {
                    } else {
                        formationChecklist = false
                    }
                }
            }
        }

        if (mOutputHatches.size > 1) {
            formationChecklist = false
        }

        if (mInputHatches.size == 1 || mInputBusses.size == 1) {
            isFluidEngine = mInputHatches.size == 1
            return formationChecklist
        }

        return false
    }

    override fun setupMachine(aBaseMetaTileEntity: IGregTechTileEntity?, stack: ItemStack?) {
        super.setupMachine(aBaseMetaTileEntity, stack)
        if (mMachine) AeroStateNetworkManager.markUpdate()
    }

    override fun addInputToMachineList(aTileEntity: IGregTechTileEntity?, aBaseCasingIndex: Int): Boolean {
        return addCargoHatch(aTileEntity, aBaseCasingIndex)
                || super.addOutputToMachineList(aTileEntity, aBaseCasingIndex)
                || super.addInputToMachineList(aTileEntity, aBaseCasingIndex)
    }

    private fun addCargoHatch(te: IGregTechTileEntity?, aBaseCasingIndex: Int): Boolean {
        if (te == null) return false
        val mte = te.metaTileEntity
        if (mte != null) {
            if (mte is GT_MetaTileEntity_Hatch) {
                mte.updateTexture(aBaseCasingIndex)
            }
            if (mte is GTMTEAerostatCargoHatch) {
                cargoHatch = mte
                return true
            }
        }
        return false
    }

    override fun onPostTick(aBaseMetaTileEntity: IGregTechTileEntity, aTick: Long) {
        super.onPostTick(aBaseMetaTileEntity, aTick)
        if (aBaseMetaTileEntity.isServerSide && mMachine && aTick % 18 == 0L) {
            if (isFluidEngine) {
                if (curBuffer < MAX_BUFFER - 1000) {
                    isFullBuffer = false
                }

                if (!isFullBuffer && depleteInput(Materials.NatruralGas.getGas(1000)) || depleteInput(Materials.Gas.getGas(1000))) {
                    curBuffer += 1000
                    if (curBuffer + 1000 > MAX_BUFFER) {
                        curBuffer = MAX_BUFFER
                        isFullBuffer = true
                    }
                }
            } else {
                val items: List<ItemStack> = storedInputs

                if (items.isNotEmpty()) {
                    val input = storedInputs[0]

                    if (input != null && input.item != null && input.stackSize > 0) {
                        for (i in 0 until input.stackSize) {
                            val item = ItemStack(input.item, 1, input.itemDamage)
                            val burnTime = TileEntityFurnace.getItemBurnTime(item)

                            if (curBuffer < MAX_BUFFER - burnTime) {
                                isFullBuffer = false
                            }

                            if (!isFullBuffer && burnTime > 0 && depleteInput(item)) {
                                input.stackSize--
                                curBuffer += burnTime
                                if (curBuffer + burnTime >= MAX_BUFFER) {
                                    curBuffer = MAX_BUFFER
                                    isFullBuffer = true
                                }
                            }
                        }
                    }
                    updateSlots()
                }
            }
        }
    }

    override fun createTooltip(): MultiBlockTooltipBuilder {
        val b = MultiBlockTooltipBuilder("aerostate")
        b
            .addInfo("info.0", "Moving over a distance by fuel")
            .addInfo("info.3", "Natural/Refinery Gas: 64 chunks or Any Burn Time Fuel: 16 chunks")
            .addInfo("info.2", "Minimum height (Y coord) for work: 60")
            .addTypeMachine("type", "Moving Machine")
            .addController()
            .sizeStructure(3, 1, 3)
            .addInputHatch(1)
            .addCasingInfo("case.1", "Aerostate Platform Casing", 7)
            .signAndFinalize()
        return b
    }

    override fun onRightclick(aBaseMetaTileEntity: IGregTechTileEntity, aPlayer: EntityPlayer): Boolean {
        return mMachine && super.onRightclick(aBaseMetaTileEntity, aPlayer)
    }

    fun setLocationName(name: String) {
        this.name = name
        impact.proxy.addClientSideChatMessages("Set location name: $name")

        if (baseMetaTileEntity.isServerSide)
            AeroStateNetworkManager.addAeroState(owner, this, uuid)
    }

    override fun onFirstTick(te: IGregTechTileEntity) {
        super.onFirstTick(te)
        if (te.isServerSide) {
            if (owner.isEmpty()) owner = te.ownerName
            if (uuid.isEmpty()) uuid = UUID.randomUUID().toString()
            if (owner.isNotEmpty() && name.isNotEmpty()) {
                AeroStateNetworkManager.addAeroState(owner, this, uuid)
            }
        }
    }

    override fun saveNBTData(aNBT: NBTTagCompound) {
        super.saveNBTData(aNBT)
        aNBT.setInteger("curBuffer", curBuffer)
        aNBT.setString("aerName", name)
        aNBT.setString("playerName", owner)
        aNBT.setString("uuidName", uuid)
        aNBT.setBoolean("firstOpen", firstOpen)
        aNBT.setBoolean("isFluidEngine", isFluidEngine)

        targetPos?.also {
            aNBT.setTag("targetPos", it.saveToNBT())
        }
    }

    override fun inValidate() {
        cargoHatch?.changeTarget(null, null)
        AeroStateNetworkManager.onRemove(owner, this, uuid)
        super.inValidate()
    }

    override fun loadNBTData(aNBT: NBTTagCompound) {
        super.loadNBTData(aNBT)
        curBuffer = aNBT.getInteger("curBuffer")
        firstOpen = aNBT.getBoolean("firstOpen")
        name = aNBT.getString("aerName")
        uuid = aNBT.getString("uuidName")
        owner = aNBT.getString("playerName")
        isFluidEngine = aNBT.getBoolean("isFluidEngine")

        aNBT.getCompoundTag("targetPos")?.also {
            val pos = PositionObject.loadFromNBT(it)
            if (!pos.isInvalid()) targetPos = pos
        }
    }

    override fun checkRecipe(itemStack: ItemStack?): Boolean {
        return false
    }

    private val distanceTravel: Int
        get() {
            val thisPos = pos
            val newPos = targetPos
            if (newPos != null && thisPos != newPos) {
                val distance = Utilits.distanceBetween3D(thisPos.x, newPos.x, thisPos.y, newPos.y, thisPos.z, newPos.z)
                if (distance >= 16) {
                    return distance
                }
            }
            return 0
        }

    fun updateSelected(indexSelect: Int) {
        this.indexSelect = indexSelect
        if (currentLocationPlatforms.isNotEmpty()) {
            val newPos = currentLocationPlatforms[indexSelect]
            targetPos = newPos.pos
            cargoHatch?.changeTarget(newPos.pos) {
                (curBuffer - selectedDistance * 25 >= 0).also { isOk ->
                    if (isOk) curBuffer -= selectedDistance * 25
                }
            }
            selectedDistance = distanceTravel
        }
    }

    fun toTravel(aPlayer: EntityPlayer) {

        if (currentLocationPlatforms.isEmpty()) {
            targetPos = null
        }

        val thisPos = pos
        targetPos?.also { newPos ->
            val distance = Utilits.distanceBetween3D(thisPos.x, newPos.x, thisPos.y, newPos.y, thisPos.z, newPos.z)
            if (distance >= 16 && thisPos.d == newPos.d) {
                if (curBuffer - distance * 25 > 0) {
                    curBuffer -= distance * 25
                    aPlayer.setPositionAndUpdate(newPos.x + 0.5, newPos.y + 1.0, newPos.z + 0.5)
                } else {
                    GT_Utility.sendChatToPlayer(aPlayer, "No Fuel")
                }
            } else {
                GT_Utility.sendChatToPlayer(aPlayer, "Error Distance")
            }
        }
    }

    override fun onScrewdriverRightClick(aSide: Byte, aPlayer: EntityPlayer, aX: Float, aY: Float, aZ: Float) {
        super.onScrewdriverRightClick(aSide, aPlayer, aX, aY, aZ)
    }

    override fun construct(itemStack: ItemStack, b: Boolean) {
        buildPiece(itemStack, b, 1, 0, 1)
    }

    override fun receive(data: ByteArrayDataInput) {
        setLocationName(data.readUTF())
    }

    override fun hasSeparate(): Boolean {
        return false
    }

    fun inputItems(inventory: Array<ItemStack?>): Boolean {
        if (mOutputBusses.isNotEmpty()) {
            val items = inventory.clone()
            Arrays.fill(inventory, null)
            for (item in items) {
                addOutput(item)
            }
            return true
        }
        return false
    }

    override val pos: IPosition
        get() = PositionObject(baseMetaTileEntity)

    override fun updateState(nested: List<IAeroStat>) {
        val te = baseMetaTileEntity ?: return
        val updateList = getRadiusAeroStates(nested, te.ownerName, te, if (isFluidEngine) 1024 else 256)
        currentLocationPlatforms.clear()
        currentLocationPlatforms.addAll(updateList)
        selectedDistance = distanceTravel

        val index = currentLocationPlatforms.indexOfFirst { targetPos == it.pos }
        if (index > 0) updateSelected(index)
    }

    private fun getRadiusAeroStates(updateList: List<IAeroStat>, owner: String, te: IGregTechTileEntity, maxRange: Int): List<IAeroStat> {
        val list: MutableList<IAeroStat> = ArrayList()
        val thisPos = PositionObject(te)

        for (aerostat in updateList) {
            val location = aerostat.pos
            val teTarget = te.getIGregTechTileEntity(location.x, location.y, location.z)
            if (teTarget == null || thisPos.isEquals(location) || thisPos.dPos != location.d) continue
            if (teTarget.metaTileEntity is GTMTE_Aerostat) {
                val gtmteAerostat = teTarget.metaTileEntity as GTMTE_Aerostat
                if (gtmteAerostat.owner == owner) {
                    val distance = Utilits.distanceBetween3D(thisPos.xPos, location.x, thisPos.yPos, location.y, thisPos.zPos, location.z)
                    if (distance in 16..maxRange) {
                        list.add(gtmteAerostat)
                    }
                }
            }
        }

        return list
    }

    override fun isValid(): Boolean {
        return baseMetaTileEntity?.metaTileEntity != null
    }
}
